### ISSUE TITLE
`bind_tf_idf` produces negative values; breaks down counts by an irrelevant variable

### ISSUE THREAD

**Issue by skolenik** [juliasilge/tidytext#99]
I am following the `bind_tf_idf()` example from the **Tidy Text** book with my data (attached; if there's any interest in how I arrived at that, I can share the data and the code up the stream, too)
[ast.zip](https://github.com/juliasilge/tidytext/files/1678323/ast.zip):

```
 ast <- readRDS("ast.Rds")
 ast %>%   count(year,word,sort=TRUE) %>% 
   ungroup() %>% 
   bind_tf_idf(word,year,n) %>%
   filter(n>4) %>%
   arrange(desc(tf_idf)) %>% 
   mutate(word=factor(word,levels=rev(unique(word)))) %>%
   filter(word=="race")
```

```
## A tibble: 4 x 7
##   year Abstract_Type     word       n       tf    idf    tf_idf
##  <int> <chr>             <fctr> <int>    <dbl>  <dbl>     <dbl>
## 1  2015 Contributed       race       5 0.000210 -0.847 -0.000178
## 2  2013 Topic Contributed race       5 0.000216 -0.847 -0.000183
## 3  2014 Contributed       race       5 0.000253 -0.847 -0.000214
## 4  2013 Contributed       race       7 0.000302 -0.847 -0.000256
```

The results break two of my expectations:

1. that `tf_idf` would be positive, and
2. that the counts would be combined over `year` (but instead, they appear to be produced by something like `count(.,year,Abstract_Type)`... this of course could be the tidyness issue as `tidytext::bind_tf_idf()` may not be expecting anything else in the data set except the tokens and the relevant grouping variable).

**Comment by EmilHvitfeldt** [comment]
So I think I have found your problem. Firstly is the `ast` already a `grouped_df` that is grouped by `year` and `Abstract_Type` hence the counting is done over the different year*Abstract_Type combinations. What you would want to do is have it ungroup before this pipeline or explicitly group_by. 

If your intension is to group by year and calculate tf_idf by year you can use the following code

``` r
library(tidyverse)
#> ── Attaching packages ───────────────────── tidyverse 1.2.1 ──
#> ✔ ggplot2 2.2.1.9000     ✔ purrr   0.2.4     
#> ✔ tibble  1.4.2          ✔ dplyr   0.7.4.9002
#> ✔ tidyr   0.8.0          ✔ stringr 1.3.0     
#> ✔ readr   1.1.1          ✔ forcats 0.2.0
#> ── Conflicts ──────────────────────── tidyverse_conflicts() ──
#> ✖ dplyr::filter() masks stats::filter()
#> ✖ dplyr::lag()    masks stats::lag()
library(tidytext)
ast <- readRDS("~/Desktop/ast.Rds")
ast %>%
  group_by(year) %>%
  count(word, sort = TRUE) %>%
  ungroup() %>% 
  bind_tf_idf(word, year, n) %>%
  arrange(desc(tf_idf)) %>% 
  mutate(word = factor(word, levels = rev(unique(word)))) %>%
  filter(word == "race")
#> # A tibble: 3 x 6
#>    year word      n       tf   idf tf_idf
#>   <int> <fct> <int>    <dbl> <dbl>  <dbl>
#> 1  2013 race     12 0.000518     0      0
#> 2  2014 race      7 0.000354     0      0
#> 3  2015 race      7 0.000294     0      0
```

The reason tf_idf is zero is because since the word "race" appears in every document, and you have that idf is defined as

<a href="https://www.codecogs.com/eqnedit.php?latex=idf(t,&space;D)&space;=&space;log\left(\frac{N}{n_t}\right)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?idf(t,&space;D)&space;=&space;log\left(\frac{N}{n_t}\right)" title="idf(t, D) = log\left(\frac{N}{n_t}\right)" /></a>

where where the idf of a given term (t) in a set of documents (D) is a function of the total number of documents being assessed (N) and the number of documents where the term t appears (n_t). Since the word "race" appears in every document we get

<a href="https://www.codecogs.com/eqnedit.php?latex=idf(t,&space;D)&space;=&space;log\left(\frac{3}{3}\right)&space;=&space;0" target="_blank"><img src="https://latex.codecogs.com/gif.latex?idf(t,&space;D)&space;=&space;log\left(\frac{3}{3}\right)&space;=&space;0" title="idf(t, D) = log\left(\frac{3}{3}\right) = 0" /></a>

tf_idf is simply the product of tf and idf, and since idf is zero so is tf_idf.

The reason why you get negative idf values is because since your data was counted inside the  year*Abstract_Type combinations but was tf_idf'ed according to year alone you got a situation where the number of documents being assessed is lower then the number of documents where the term appears. Because N is calculated as number of unique values of year and n_t as the number of rows in the data.frame with the word "race" in it.

``` r
ast %>%
  filter(word == "race") %>%
  group_by(year, Abstract_Type) %>%
  count(word, sort = TRUE)
#> # A tibble: 7 x 4
#> # Groups:   year, Abstract_Type [7]
#>    year Abstract_Type     word      n
#>   <int> <chr>             <chr> <int>
#> 1  2013 Contributed       race      7
#> 2  2013 Topic Contributed race      5
#> 3  2014 Contributed       race      5
#> 4  2015 Contributed       race      5
#> 5  2015 Topic Contributed race      2
#> 6  2014 Invited           race      1
#> 7  2014 Topic Contributed race      1
```
So it thought that you have the word in 7 out of 3 documents which gives you a idf of

<a href="https://www.codecogs.com/eqnedit.php?latex=idf(t,&space;D)&space;=&space;log\left(\frac{3}{7}\right)&space;=&space;-0.8472979" target="_blank"><img src="https://latex.codecogs.com/gif.latex?idf(t,&space;D)&space;=&space;log\left(\frac{3}{7}\right)&space;=&space;-0.8472979" title="idf(t, D) = log\left(\frac{3}{7}\right) = -0.8472979" /></a>

Which was what you got too.

**Comment by juliasilge** [comment]
This has now been addressed in #112 thanks to @EmilHvitfeldt. I'm so glad @skolenik brought this to our attention and that this has been merged now! Thanks to all. 🎉
